CPP11 message serialisation aliasing fixed

hdino opened the pull request:
This patch fixes the aliasing bug mentioned in #365.

@tridge Please review.

vooon commented on generator/CPP11/include_v2.0/msgmap.hpp:
```diff
@@ -39,7 +39,7 @@ class MsgMap {
 
 	template<typename _T>
-	void operator<< (const _T data);
+	void operator<< (const _T &data);
```
What is the reason to pass by ref?

hdino replied:
It avoids a copy (though there won't be any benefit for data types <= pointer size).

vooon replied:
It even can give speed penalty because you have to access pointer.

vooon commented on generator/CPP11/include_v2.0/msgmap.hpp:
```diff
@@ -66,70 +63,47 @@ class MsgMap {
-		buf = *static_cast<const _T *>(static_cast<const void *>(&data));
-	} else {
-		buf = data;
-	}
```
Actually you can just put `memcpy()` instead of that deref.

hdino replied:
Indeed, that's true. I changed the design, because previously both `MsgMap::msg_swap_memcpy` and `MsgMap::operator<<` modified `_MAV_PAYLOAD_NON_CONST` (see [here](https://github.com/ArduPilot/pymavlink/blob/master/generator/CPP11/include_v2.0/msgmap.hpp#L97) and [here](https://github.com/ArduPilot/pymavlink/blob/master/generator/CPP11/include_v2.0/msgmap.hpp#L108)).

vooon replied:
`msg_swap_memcpy()` is just a way to avoid copy-paste in `<<`.

hdino commented:
I created a new version that avoids the switch statement and is easier to read.

vooon commented:
@hdino yes, that looks better.
Need to do same for `operator>>` (from le to host). But please note that in mavlink v2 message with trailing zero bytes may be trimmed.

hdino commented:
@vooon I refactored the code for `operator>>`, please have a look.

To keep things consistent, I also adjusted some of the previous changes.

Should we move the newly introduced functions into the `mavlink` namespace?

vooon commented on generator/CPP11/include_v2.0/msgmap.hpp:
```diff
@@ -140,94 +155,78 @@ void mavlink::MsgMap::operator<< (const std::array<_T, _Size> &data)
-	case 1:
-		data = _MAV_PAYLOAD(cmsg)[pos];
-		break;
+    int remaining_non_zero_data = cmsg->len - pos;
```
Probably better to use ssize_t.

vooon commented:
Please fix: https://travis-ci.org/ArduPilot/pymavlink/jobs/615053828#L8701

https://en.cppreference.com/w/cpp/language/static_assert - add message.

hdino commented:
@vooon Thanks for your suggestions! Before merging, please check if the padding with zeroes is still correct, because I implemented it according to the documentation.

hdino commented on generator/CPP11/include_v2.0/msgmap.hpp:
```diff
@@ -54,180 +55,185 @@ class MsgMap {
+    ssize_t remaining_non_zero_data = cmsg->len - pos;
+    typename mavlink::impl::UintBuffer<_T>::Type buf;
+
+    if (static_cast<ssize_t>(sizeof(_T)) <= remaining_non_zero_data) { // field is not truncated
```
I changed this cast to `ssize_t`

hdino commented on generator/CPP11/include_v2.0/msgmap.hpp:
```diff
@@ -54,180 +55,185 @@ class MsgMap {
+        memcpy(&buf, &_MAV_PAYLOAD(cmsg)[pos], sizeof(_T));
+    } else if (remaining_non_zero_data <= 0) {
+        // there is no non-zero data left, so just fill with 0
+        buf = 0;
```
Added this optimisation for the case when there's no data left.
